sycl/doc/extensions/proposed/sycl_ext_oneapi_joint_for.asciidoc

```diff
@@ -0,0 +1,210 @@
+
+[source,c++]
+----
+for (int i = it.get_local_id(0); i < N; i += it.get_local_range(0)) { ... }
+----
+
+The combination of setting the initial value based on the work-item index and
+setting the stride based on the number of work-items in the group ensures that
+any data accesses within the loop are appropriately interleaved to maximize
+performance.
+
+Although common, this idiom has several drawbacks: it is difficult to teach to
+developers unfamiliar with SPMD programming; it is easy to use incorrectly
+(e.g., by forgetting to set a different initial value on each work-item); and
+it is difficult to use generically (i.e., for different groups of work-items).
+
+This extension proposes to encapsulate this pattern in group algorithms, to
+simplify generic usage across different group types.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_JOINT_FOR` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Group algorithms
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template <typename Group, typename InputIterator, typename Function>
+Function joint_for_each(Group g, InputIterator first, InputIterator last, Function f);
+
+} // namespace sycl::ext::oneapi
+----
+
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>>` is
+true and `InputIterator` is a random access iterator.
+
+_Preconditions_: `first` and `last` must be the same for all work-items in
+group `g`, and `f` must be an immutable callable with the same type and state
+for all work-items in group `g`.
+
+_Effects_: Blocks until all work-items in group `g` have
+reached this synchronization point, then applies `f` to the result of
+dereferencing every iterator in the range [`first`, `last`).
+The range is distributed across the work-items in group `g`.
+
+_Synchronization_: The call to this function in each work-item
+happens before the algorithm begins execution.
+The completion of the algorithm happens before any work-item
+blocking on the same synchronization point is unblocked.
+
+_Remarks_: If `f` returns a result, the result is ignored.
+If the range [`first`, `last`) is not evenly divisible by the number of
+work-items in group `g`, the number of times that `f` is invoked will differ
+across work-items.
+
+_Returns_: `f`.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template <typename Group, typename Integer, typename Function>
+Function joint_for(Group g, Integer first, Integer last, Function f);
```

Comment:
I'm just curious.  Is this a common operation to iterate over a sequence of integers?

Reply:
The idea of this function was to replace this common idiom:

```c++
for (int i = g.get_local_linear_id() + first; i < last; i += g.get_local_linear_range()) {
  f(i);
}
```

with:

```c++
sycl::joint_for(g, first, last, f);
```

which avoids all the complexity of creating something like a `counting_iterator` just to use `for_each`.